packages/cli-repl/src/logger.ts

```diff
@@ -1,53 +1,189 @@
 import redactInfo from 'mongodb-redact';
+import Analytics from 'analytics-node';
 import redactPwd from './redact-pwd';
 import { ObjectId } from 'bson';
 import pino from 'pino';
-import clr from './clr';
 import path from 'path';
 import os from 'os';
 
-function logger(bus: any, logDir: string) {
+interface ApiEventArguments {
+  pipeline?: any[];
+  query?: object;
+  options?: object;
+  filter?: object; 
+}
+
+interface ApiEvent {
+  method?: string; 
+  class?: string;
+  db?: string;
+  coll?: string;
+  arguments?: ApiEventArguments;
+}
+
+interface UseEvent {
+  db: string;
+}
+
+interface AsyncRewriterEvent {
+  original: string;
+  rewritten: string;
+}
+
+interface ShowEvent {
+  method: string;
+}
+
+interface ErrorEvent {
+  error: object;
+}
+
+interface ConnectEvent {
+  uri: string;
+  isAtlas: boolean;
+  isLocalhost: boolean;
+  serverVersion: string;
+  isEnterprise: boolean;
+  authType: string;
+  isDataLake: boolean;
+  dlVersion?: string;
+  isGenuine: boolean;
+  serverName: string;
+}
+
+export default function logger(bus: any, logDir: string) {
   const sessionID = new ObjectId(Date.now());
   const logDest = path.join(logDir, `${sessionID}_log`);
   const log = pino({ name: 'monogsh' }, pino.destination(logDest));
   console.log(`Current sessionID: ${sessionID}`);
+  let userId;
+  let telemetry;
+
+  let analytics = new NoopAnalytics();
+  try {
+    // this file gets written as a part of a release
+    console.log(require('./config.js'))
+    log.warn(require('./config.js').SEGMENT_API_KEY)
+    analytics = new Analytics(require('./config.js').SEGMENT_API_KEY);
+  } catch (e) {
+    bus.emit('mongosh:error', e)
+  }
 
-  bus.on('mongosh:connect', function(info) {
-    const params = { sessionID, info: redactPwd(info) };
-    log.info('connect', params);
+  bus.on('mongosh:connect', function(args: ConnectEvent) {
+    const connectionUri = redactPwd(args.uri);
+    delete args.uri;
+    const params = { sessionID, userId, connectionUri, ...args };
+    log.info('mongosh:connect', params);
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:connect',
+        properties: { sessionID, connectionUri, ...args }
+      });
+    }
   });
 
-  bus.on('mongosh:error', function(error) {
+  bus.on('mongosh:new-user', function(id, enableTelemetry) {
+    userId = id;
+    telemetry = enableTelemetry;
+    if (telemetry) analytics.identify({ userId });
+  })
+
+  bus.on('mongosh:update-user', function(id, enableTelemetry) {
+    userId = id;
+    telemetry = enableTelemetry;
+    log.info('mongosh:update-user', { enableTelemetry })
+  })
+
+
+  bus.on('mongosh:error', function(error: ErrorEvent) {
     log.error(error);
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:error',
+        properties: { error }
```

Comment:
One of the side effects of how the node repl handles errors weirdly, is that we can send errors that we've thrown to telemetry but skip other errors, right? Since we don't want to send every JS error ever triggered on the shell to telemetry? I'm not 100% clear on what errors are getting sent

Reply:
good question! so we currently send everything, because all the errors ever make it to the writer [get logged](https://github.com/mongodb-js/mongosh/blob/master/packages/cli-repl/src/cli-repl.ts#L202). This doesn't include unhandled exceptions, since we don't currently handle those. 

Do you think we should be a bit more specific about what errors we send back? For example, only `MongoshInternalError`? I do agree, I think telemetry will be super polluted if we do everything.

Reply:
Yeah I'd say even only Mongosh errors and not generic javascript errors that get written to the logger but not thrown by us.

Reply:
ok, now telemetry only gets Mongosh errors. Local logs get everything tho, because I want to keep an eye on all the js throws in case we have missed something.

Reply:
Perfect, ty!